### PR TITLE
feat(api): USCIS H-1B Employer Data Hub import, h1b_sponsors table, and sponsor_likelihood matching (#262)

### LINKED ISSUE #262: feat(api): USCIS H-1B Employer Data Hub import + h1b_sponsors table + sponsor_likelihood matching
> **Key:** `#262` · **Size:** M · **Epic:** #245 · **Program:** #243 · **Blocked by:** #258
>
> ⚠️ **Migration numbers in this ticket are nominal.** Parity epics were drafted in parallel; at branch time check `apps/api/db/migrations/` and use the actual next free number (keep the descriptive suffix).

## Context

Spec §2 locks "US + sponsorship signal (USCIS H-1B employer data)"; §7: *"normalized employer match against `h1b_sponsors` → `sponsor_likelihood` + approval counts on the job card; feed filter 'likely sponsors only'. Data refresh = annual script run."* §5 defines the table: *"`h1b_sponsors` — normalized `employer_name`, `fiscal_year`, `approvals`, `denials` (USCIS H-1B Employer Data Hub import; annual refresh script)."*

The USCIS H-1B Employer Data Hub publishes annual CSV exports (uscis.gov → Tools → H-1B Employer Data Hub → Files). Relevant columns (2020s-era exports): `Fiscal Year`, `Employer (Petitioner) Name`, `Initial Approval`, `Initial Denial`, `Continuing Approval`, `Continuing Denial` (plus NAICS/state columns we ignore). One employer can appear on many rows (one per state/NAICS slice) — the import must aggregate per (employer, year).

#258 added the `jobs.sponsor_likelihood jsonb` column (shape `{"status":"known_sponsor","approvals":N,"denials":N}`, null = unknown). This ticket populates it: an import script, a matcher lib, and wiring into discovery inserts. Scripts in `apps/api` run via `tsx` (see `apps/api/package.json`'s `db:init` script and `apps/api/scripts/db-init.ts` for the pattern of a script that opens a pg pool from `DATABASE_URL`).

## Scope

- Migration for `h1b_sponsors`; employer-name normalization + lookup lib; import script (CSV → upserts) + post-import backfill of existing jobs; discovery enrichment sets `sponsor_likelihood` on insert.

**Out of scope:** the feed filter UI (#268); automatic annual refresh scheduling (the run is manual-by-design: download CSV, run script).

## Implementation guide

1. **Create** `db/migrations/014_h1b_sponsors.sql` (next free number at branch time):

```sql
-- USCIS H-1B Employer Data Hub snapshot (Jobright-parity Epic 2, spec §5/§7).
-- Global reference data (no user_id): one row per (employer, fiscal year),
-- aggregated across the per-state/NAICS rows of the USCIS export.
create table if not exists h1b_sponsors (
  id bigint generated always as identity primary key,
  employer_name text not null,
  employer_name_normalized text not null,
  fiscal_year integer not null,
  approvals integer not null default 0,
  denials integer not null default 0,
  created_at timestamptz not null default now()
);

create unique index if not exists h1b_sponsors_employer_year_unique_idx
  on h1b_sponsors (employer_name_normalized, fiscal_year);
create index if not exists h1b_sponsors_normalized_idx
  on h1b_sponsors (employer_name_normalized);
```

2. **Create** `apps/api/src/lib/sponsorship.ts`:

```ts
import { getPool } from '@/lib/postgres';
import type { SponsorLikelihood } from '@/types';

const CORPORATE_SUFFIXES = [
  'incorporated', 'corporation', 'company', 'holdings', 'group',
  'inc', 'llc', 'llp', 'lp', 'plc', 'pllc', 'ltd', 'corp', 'co', 'usa', 'us',
];

/**
 * Canonical employer key: uppercase-insensitive, punctuation-free, corporate
 * suffixes stripped from the END repeatedly ("Acme Holdings LLC" -> "acme"),
 * whitespace collapsed. Never strips a suffix that would empty the name
 * ("CO Inc" -> "co").
 */
export function normalizeEmployerName(name: string): string {
  let words = name
    .toLowerCase()
    .replace(/&/g, ' and ')
    .replace(/[^a-z0-9 ]+/g, ' ')
    .split(/\s+/)
    .filter(Boolean);
  while (words.length > 1 && CORPORATE_SUFFIXES.includes(words[words.length - 1]!)) {
    words = words.slice(0, -1);
  }
  return words.join(' ');
}

/** How many trailing fiscal years count toward the snapshot. */
const LOOKBACK_YEARS = 3;

/**
 * Aggregate approvals/denials for a company over the last LOOKBACK_YEARS fiscal
 * years present in h1b_sponsors. Exact normalized-name match only (fail closed:
 * no fuzzy matching — a wrong "known sponsor" claim is worse than "unknown").
 * Returns null (unknown) when Postgres is unavailable or no row matches.
 */
export async function lookupSponsorLikelihood(company: string): Promise<SponsorLikelihood | null> {
  const pool = getPool();
  if (!pool) return null;
  const key = normalizeEmployerName(company);
  if (!key) return null;
  try {
    const { rows } = await pool.query<{ approvals: string; denials: string }>(
      `select coalesce(sum(approvals), 0) as approvals, coalesce(sum(denials), 0) as denials
       from h1b_sponsors
       where employer_name_normalized = $1
         and fiscal_year >= (select coalesce(max(fiscal_year), 0) from h1b_sponsors) - $2`,
      [key, LOOKBACK_YEARS - 1],
    );
    const approvals = Number(rows[0]?.approvals ?? 0);
    const denials = Number(rows[0]?.denials ?? 0);
    if (approvals + denials === 0) return null;
    return { status: 'known_sponsor', approvals, denials };
  } catch {
    return null; // lookup is best-effort enrichment, never blocks an insert
  }
}
```

3. **Add dependency**: `csv-parse` (^6) to `apps/api` **devDependencies** (script-only; `npm install -w @jobops/api --save-dev csv-parse`).

4. **Create** `apps/api/scripts/import-h1b.ts` — run as `npx tsx apps/api/scripts/import-h1b.ts <path-to-uscis-export.csv>` with `DATABASE_URL` set:
- Parse with `import { parse } from 'csv-parse/sync';` using `{ columns: true, bom: true, skip_empty_lines: true, relax_column_count: true }`.
- Column access must tolerate header drift across USCIS years: find columns case-insensitively — fiscal year header matching `/fiscal year/i`, employer matching `/employer.*name|petitioner.*name/i`, and the four count columns `/initial approval/i`, `/initial denial/i`, `/continuing approval/i`, `/continuing denial/i`. Counts arrive as strings possibly with commas or blank → `Number(String(v).replace(/,/g, '')) || 0`.
- Aggregate in-memory into a `Map<string /* norm|year */, { name, year, approvals, denials }>` where `approvals = initialApproval + continuingApproval` and `denials = initialDenial + continuingDenial`.
- Upsert in batches of 500:

```sql
insert into h1b_sponsors (employer_name, employer_name_normalized, fiscal_year, approvals, denials)
values ($1, $2, $3, $4, $5)
on conflict (employer_name_normalized, fiscal_year)
do update set approvals = excluded.approvals, denials = excluded.denials, employer_name = excluded.employer_name
```

- Then backfill existing jobs in one statement and print the count:

```sql
update jobs j
set sponsor_likelihood = jsonb_build_object('status', 'known_sponsor', 'approvals', s.approvals, 'denials', s.denials)
from (
  select employer_name_normalized, sum(approvals) as approvals, sum(denials) as denials
  from h1b_sponsors
  where fiscal_year >= (select max(fiscal_year) from h1b_sponsors) - 2
  group by employer_name_normalized
) s
where s.employer_name_normalized = lower(regexp_replace(j.company, '[^a-zA-Z0-9 ]+', ' ', 'g'))
```

(Exact-ish SQL-side normalization can't replicate suffix-stripping; simpler and correct: select all distinct `jobs.company`, normalize in TypeScript with `normalizeEmployerName`, look up each, and issue per-company `update jobs set sponsor_likelihood = $1 where company = $2` — do it this way, it reuses the one canonical normalizer. Log `updated N jobs across M companies`.)
- Exit non-zero with a clear message on missing file/columns/DATABASE_URL.

5. **Wire into discovery**: in `apps/api/src/lib/discovery.ts`, add dep `lookupSponsor: typeof lookupSponsorLikelihood` (default in `routes/discovery.ts`), and inside the shared insert helper, before `createJob`: `const sponsor = await deps.lookupSponsor(job.company);` then persist it. Since `createJob` (#258) doesn't take `sponsorLikelihood`, extend `CreateJobBody` with `sponsorLikelihood?: SponsorLikelihood | null` and pass it through `createJob`'s INSERT (`sponsor_likelihood` column, `JSON.stringify` via pg jsonb param). This keeps manual job creation unaffected (undefined → null).

## Testing

Run: `npm test -w @jobops/api` (the test runner also globs `apps/api/scripts/**/*.test.ts` — see `apps/api/scripts/run-tests.mjs`).

**Create** `apps/api/src/lib/sponsorship.test.ts` — spec §11 explicitly calls for *"sponsorship matcher (name-normalization edge cases)"*:

```ts
import { strict as assert } from 'node:assert';
import { describe, it } from 'node:test';
import { normalizeEmployerName } from './sponsorship';

describe('normalizeEmployerName', () => {
  it('strips punctuation and one suffix', () => {
    assert.equal(normalizeEmployerName('Google, LLC'), 'google');
  });
  it('strips stacked suffixes', () => {
    assert.equal(normalizeEmployerName('Acme Holdings LLC'), 'acme');
  });
  it('maps & to and', () => {
    assert.equal(normalizeEmployerName('Ernst & Young U.S. LLP'), 'ernst and young');
  });
  it('never empties a name made of suffix words', () => {
    assert.equal(normalizeEmployerName('CO Inc'), 'co');
  });
  it('collapses whitespace and case', () => {
    assert.equal(normalizeEmployerName('  MICROSOFT   CORPORATION '), 'microsoft');
  });
});
```

**Create** `apps/api/scripts/import-h1b.test.ts` for the pure pieces (export the row-aggregation + header-resolution functions from the script): a 4-row fixture (same employer, two states, two years, comma-formatted counts) aggregates to 2 upsert rows with summed counts; a fixture missing the employer column throws with a message naming the column.

Manual verification (documented in the script's header comment): download one Data Hub CSV, run the script locally against the dev DB, then `select * from h1b_sponsors order by approvals desc limit 5;`.

## Acceptance criteria
- [ ] Migration applies + re-applies cleanly.
- [ ] Import script ingests a real USCIS export end-to-end (aggregated per employer/year, re-runnable — second run updates, not duplicates).
- [ ] After import, existing jobs whose company matches get `sponsor_likelihood` backfilled; newly discovered jobs get it at insert.
- [ ] Matching is exact-on-normalized-name only; no match → `null` (unknown), never a guessed "known_sponsor".
- [ ] `npm test -w @jobops/api` + typecheck + lint pass.

## Dependencies
Blocked by: #258 (`sponsor_likelihood` column). Unblocks: #268 (sponsorship filter/badges), #265 (sponsorship sub-signal input).

## Invariants
- Spec §12.2 (fail closed): a lookup failure or no-match yields `unknown` (null), never a fabricated sponsorship claim.
- Spec §12.3: the import reads a **local file** the owner downloaded; the script performs no network fetches at all.

### PR DESCRIPTION
**Pull request by Taleef7**
Closes #262 (Epic #245 — Feed Engine).

### Summary of Changes

1. **Database Migration (`db/migrations/015_h1b_sponsors.sql`)**:
   - Creates global reference table `h1b_sponsors` with columns `id`, `employer_name`, `employer_name_normalized`, `fiscal_year`, `approvals`, `denials`, and `created_at`.
   - Adds unique index on `(employer_name_normalized, fiscal_year)` and lookup index on `(employer_name_normalized)`.
   - Alters `jobs.sponsor_likelihood` to `text` to accommodate JSON-serialized `{ status: 'known_sponsor', approvals, denials }` payloads.

2. **Employer Normalization & Sponsorship Lookup (`apps/api/src/lib/sponsorship.ts`)**:
   - `normalizeEmployerName(name)`: strips punctuation, maps `&` to `and`, normalizes whitespace/case, and repeatedly strips stacked trailing corporate suffixes (`llc`, `inc`, `corp`, `holdings`, `group`, etc.) while protecting suffix-only names (e.g. `'CO Inc'` -> `'co'`).
   - `lookupSponsorLikelihood(company)`: exact normalized lookup over the last 3 fiscal years present in `h1b_sponsors` (`(select coalesce(max(fiscal_year), 0) from h1b_sponsors) - 2`), returning `{ status: 'known_sponsor', approvals, denials }` when positive filings exist. Fails closed (returns `null`) on missing pool, empty string, zero filings, or query error.

3. **USCIS H-1B CSV Import Script (`apps/api/scripts/import-h1b.ts`)**:
   - Local CSV parsing with `csv-parse/sync` (BOM and relax column count handling).
   - Case-insensitive header matching for fiscal year, employer name, and approval/denial counts.
   - Aggregates multi-state rows per `(employerNameNormalized, fiscalYear)`.
   - Batch upsert into `h1b_sponsors` (500 records per batch) with `ON CONFLICT DO UPDATE`.
   - Backfills existing `jobs` with `sponsor_likelihood` JSON payloads for matching companies.

4. **Types & Job Stores Integration**:
   - Extended `SponsorLikelihood` in `apps/api/src/types.ts` and `apps/web/src/types/job.ts` with `KnownSponsorLikelihood`.
   - Updated `job-store.postgres.ts` and in-memory store to serialize `sponsorLikelihood` on `createJob` and safely parse it on read.
   - In `apps/api/src/lib/discovery.ts`, `insertIfNew` enriches jobs with `deps.lookupSponsor(job.company)` and preserves pre-existing signals.
   - Wired `lookupSponsor: lookupSponsorLikelihood` into `routes/discovery.ts`.

5. **Testing & Verification**:
   - Unit tests in `sponsorship.test.ts` (normalization edge cases + mock pool queries) and `import-h1b.test.ts` (header resolution + row aggregation).
   - Integration test in `discovery.test.ts` verifying `lookupSponsor` enrichment.
   - 336 API unit tests pass. Full `npm run check` (monorepo lint + typecheck + Next.js build + API build) clean.

### CODE REVIEW

**Comment by Taleef7** [comment]
@codex review

File: apps/api/src/data/job-store.postgres.ts [hunk 1]
```diff
@@ -286,7 +295,12 @@ export async function createJob(userId: string, body: CreateJobBody): Promise<Jo
     });
   const lastSeenAt = body.lastSeenAt ?? timestamp;
   const liveness = body.liveness ?? 'active';
-  const sponsorLikelihood = body.sponsorLikelihood ?? null;
+  const sponsorLikelihood =
+    body.sponsorLikelihood !== undefined
+      ? body.sponsorLikelihood
+        ? JSON.stringify(body.sponsorLikelihood)
```
[the remark this answers is not in the document]
Comment (Taleef7) [reply apps/api/src/data/job-store.postgres.ts:304]: Fixed in commit d5c35e7. Updated createJob in job-store.postgres.ts so sponsorLikelihood is stringified only when it is an object (	ypeof body.sponsorLikelihood === 'object' ? JSON.stringify(...) : body.sponsorLikelihood), preserving unquoted strings for legacy categorical values so index queries against idx_jobs_sponsor_likelihood match consistently. Also updated parseSponsorLikelihood to conditionally parse only strings starting with {.

File: apps/api/src/types.ts [hunk 2]
```diff
@@ -16,7 +16,18 @@ export type JobPriority = 'high' | 'medium' | 'low';
...
+  approvals: number;
+  denials: number;
+}
+
+export type SponsorLikelihood =
+  | 'likely'
+  | 'possible'
+  | 'unlikely'
+  | 'unknown'
+  | KnownSponsorLikelihood;
```
[the remark this answers is not in the document]
Comment (Taleef7) [reply apps/api/src/types.ts:30]: Fixed in commit d5c35e7. Updated POST /api/jobs validation in apps/api/src/routes/jobs.ts to support both legacy status strings (likely, possible, unlikely, unknown) and { status: "known_sponsor", approvals: number, denials: number } objects with numeric validation. Added a unit test in jobs.test.ts verifying that POST /api/jobs accepts and round-trips a KnownSponsorLikelihood object payload.
